compiler/next/include/chpl/Queries/Context.h

```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_QUERIES_CONTEXT_H
+#define CHPL_QUERIES_CONTEXT_H
+
+#include "chpl/AST/UniqueString.h"
+#include "chpl/AST/ID.h"
+#include "chpl/Queries/ContextDetail.h"
+#include "chpl/Util/memory.h"
+
+#include <unordered_map>
+
+namespace chpl {
+
+/**
+
+\rst
+
+This class stores the compilation-wide context. Another name for this
+compilation-wide context is *program database*. It handles unique'd strings and
+also stores the results of queries (so that they are are memoized). It tracks
+dependencies of queries in order to update them appropriately when a dependency
+changes.
+
+Queries are functions that are written in a stylized manner to interact with the
+context (aka program database). For example, a ``parse`` query might accept as
+an argument a ``UniqueString path`` and return a vector of owned AST nodes.
+Another example is a query to determine the location of an AST node; it would
+accept as an argument a ``BaseAST*`` and it would return a Location.
+
+When running a query, the query system will manage:
+ * checking to see if the query result is already saved and available for reuse
+ * recording the queries called by that query as dependencies
+
+To write a query, create a function that uses the ``QUERY_`` macros defined in
+QueryImpl.h. The arguments to the function need to be efficient to copy (so
+``UniqueString``, ``ID``, ``Location``, and pointers are OK, but e.g.
+``std::vector`` is not).  The function will return a result, which need not be
+POD and can include AST pointers (but see below). The function needs to be
+written in a stylized way to interact with the context.
+
+For example, here is a query that computes MyResultType from myArg1 and
+myArg2:
+
+.. code-block:: c++
+
+    #include "chpl/Queries/QueryImpl.h"
+
+    const MyResultType& myQueryFunction(Context* context,
```

Comment:
Would having context as an argument make things arduous? If we only have one context (seems like it?), can we avoid having that as an argument? Maybe having `context` as a global singleton?

Reply:
I'm trying to avoid all global variables in this effort. I don't think that a single argument Context argument here makes developing these queries particularly hard. I expect someone writing a query will start with an example since there is other boiler-plate-y code in it to do the query macros.